enacts/calc.py

```diff
@@ -619,6 +620,112 @@ def daily_tobegroupedby_season(
+    end_month : int
+        Day of the end date of the season.
+    time_dim : str, optional
+        Time coordinate in `daily_data` (default `time_dim`="T").
+    Returns
+    -------  
+    See Also
+    --------
+    Notes
+    -----
+    Examples
+    --------
+    """
+    if (
+        (start_day == 29)
+        and (start_month == 2)
+        and ((~daily_data[time_dim].dt.is_leap_year).sum() > 0)
+    ) :
+        raise Exception(
+            "if there is at least one non-leap year in time_coord, can not start on 29-Feb"
+        )
+    # Have edges start early enough, and end late enough, 
+    if start_year is None :
+        year_offset = 1
+        if daily_data[time_dim][0].dt.year == daily_data[time_dim][1].dt.year :
+            year_offset = 0
+        start_year = (daily_data[time_dim][0].dt.year - year_offset).values
+    if end_year is None :
+        year_offset = 1
+        if daily_data[time_dim][0].dt.year == daily_data[time_dim][1].dt.year :
+            year_offset = 0
+        end_year = (daily_data[time_dim][-1].dt.year + year_offset).values
+    # ending season on 29-Feb means ending on 1-Mar with -1 day offset
+    if (end_day == 29 and end_month == 2) :
+        end_day = 1
+        end_month = 3
+        day_offset = pd.Timedelta(days=1)
+    else:
+        day_offset = pd.Timedelta(days=0)
+    # seasons overlapping 2 years call for 1 year offset
+    year_offset = 0
+    if (
+        datetime.datetime(start_year, start_month, start_day)
+        > datetime.datetime(start_year, end_month, end_day)
+     ) :
+        year_offset = 1
+    # Create start_/end_edges pairing arrays
+    start_edges = pd.date_range(
+        start=datetime.datetime(start_year, start_month, 1),
+        end=datetime.datetime(end_year - year_offset, start_month, 1),
+        freq="12MS",
+    ) + pd.Timedelta(days=start_day-1)
+    end_edges = pd.date_range(
+        start=datetime.datetime(start_year + year_offset, end_month, 1),
+        end=datetime.datetime(end_year, end_month, 1),
+        freq="12MS",
+    ) + pd.Timedelta(days=end_day-1) - day_offset
+    # dates mapped to the season start and end
+    season_start = xr.DataArray(np.piecewise(
+        daily_data[time_dim],
+        [((daily_data[time_dim] >= start_edges[t]) & (daily_data[time_dim] <= end_edges[t])).values
+            for t in range(start_edges.size)],
+        [start_edges[t] for t in range(start_edges.size)]
+            + [pd.to_datetime([np.nan])],
+    ), dims=[time_dim], coords={time_dim: daily_data[time_dim]})
+    season_end = xr.DataArray(np.piecewise(
+        daily_data[time_dim],
+        [((daily_data[time_dim] >= start_edges[t]) & (daily_data[time_dim] <= end_edges[t])).values
+            for t in range(end_edges.size)],
+        [end_edges[t] for t in range(end_edges.size)]
+            + [pd.to_datetime([np.nan])],
+    ), dims=[time_dim], coords={time_dim: daily_data[time_dim]})
+    # Drop days out of seasons and assign coords
```

Comment:
I am thinking that it may make sense to factor out the generation of the season_start and season_end for other potential purposes. But we might still want to keep this one specifically for groupby use because it also drops the out-of-seasons days, and I believe groupby expectes the DataArray and the group Parameter to be same size.